[DKHB0207] iP

DKHB0207 opened the pull request:
# Snipe

> "Your mind is for having ideas, not holding them." – David Allen [source](https://dansilvestre.com/productivity-quotes/)

**Snipe** frees your mind from having to remember things you need to do. It's:

- Text-based
- Easy to learn
- *Super fast* to use

All you need to do is:

1. Download it from [here](https://example.com).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you. 😊

And it is **FREE**!

## Features

- [x] Managing tasks  
- [ ] Managing deadlines (coming soon)  
- [ ] Reminders (coming soon)  

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

DKHB0207 committed:
clone Duke repo

DKHB0207 committed:
add echo and bye commands

DKHB0207 committed:
add list function

DKHB0207 committed:
add mark and unmark tasks

DKHB0207 committed:
change implementation of tasks to new class Task

DKHB0207 committed:
add ToDo, Deadline, Event to types of tasks

DKHB0207 committed:
Add test files and refine kinks in chat output

DKHB0207 committed:
Remove logo and fixed test cases

DKHB0207 committed:
Add SnipeException Class

DKHB0207 committed:
Add error handling and instructions for commands

DKHB0207 committed:
Add delete and reformat functions

DKHB0207 committed:
Add enum

DKHB0207 committed:
Add small change to output

DKHB0207 committed:
Add javadoc comments

DKHB0207 committed:
change style

DKHB0207 committed:
Add storage class

DKHB0207 committed:
Implement storage class

DKHB0207 committed:
Merge branch-Level-7 to master

DKHB0207 committed:
Add date time formats

DKHB0207 committed:
Merge branch 'branch-Level-8'

DKHB0207 committed:
Create new branch and add Parser, TaskList and Ui classes

DKHB0207 committed:
Refactor code to align to OOP principles

DKHB0207 committed:
Merge branch 'branch-A-MoreOOP'

DKHB0207 committed:
Organise classes into packages and removed unused variables

DKHB0207 committed:
Merge branch 'branch-A-Packages'

DKHB0207 committed:
Merge commit '81a9c537afe8492675eb253968939364d65db091'

DKHB0207 committed:
Add JUnit tests and fixed minor output bugs

DKHB0207 committed:
Package the App as a JAR file

DKHB0207 committed:
Add JavaDoc comments for classes in core package

DKHB0207 committed:
Add JavaDoc comments for classes in command package

DKHB0207 committed:
Add JavaDoc comments for classes in storage package

DKHB0207 committed:
Add JavaDoc comments for classes in task package

DKHB0207 committed:
Add JavaDoc comments for classes in util package

DKHB0207 committed:
Refactor the code to comply with a coding standard

DKHB0207 committed:
Add and integrate Find class into chatbot

DKHB0207 committed:
Merge branch 'branch-A-JavaDoc'

DKHB0207 committed:
Merge branch 'branch-A-CodingStandard'

DKHB0207 committed:
Add JavaDoc comments for new class FindCommand and new methods

DKHB0207 committed:
Merge branch 'branch-Level-9'

sayglenn commented on src/main/java/snipe/core/TaskList.java:
```diff
@@ -0,0 +1,111 @@
+     *
+     * @return The size of the task list.
+     */
+    public int size() {
```
Perhaps using a name like getTaskCount would be more appropriate as methods should be phrased as verbs?

sayglenn commented on src/test/java/snipe/core/SnipeTest.java:
Are these tests correct? Looks like you might be implementing them in the future, looking forward to it!

sayglenn commented on src/main/java/snipe/task/Task.java:
```diff
@@ -0,0 +1,93 @@
+     *
+     * @return A string representing the task's status.
+     */
+    public String stringStatus() {
```
This could be changed to an action, such as getStringStatus.

sayglenn commented on src/main/java/snipe/command/AddCommand.java:
```diff
@@ -0,0 +1,44 @@
+ * This command adds the specified task, saves the updated task list to storage, and
+ * provides feedback to the user via the user interface.
+ */
+public class AddCommand extends Command{
```
nit: missing space between Command and curly brace.

sayglenn commented on text-ui-test/input.txt:
Perhaps this file could be hidden and only created when user uses the app? If not, the user would have existing tasks on their database already.

sayglenn commented on src/main/java/snipe/storage/Storage.java:
```diff
@@ -0,0 +1,129 @@
+     *
+     * @throws IOException If an input or output error occurs while creating the directory or file.
+     */
+    public void initialise() throws IOException {
```
Perhaps the name of this method could be more specific, e.g. initialiseDatabase?

sayglenn reviewed (commented):
Hey Daryl! In general, your code looks good and your functions look robust. However, most of your Command classes are missing a space between `... extends Command` and the opening curly brace. I've only commented on the first occurrence of it to reduce the amount of repeated comments, but do take a look at all of the Command files.

Overall, the code is well done and coding standard is generally adhered to. Good job!

DKHB0207 committed:
Add dependencies to build.gradle

DKHB0207 committed:
Add gui packages and classes

DKHB0207 committed:
Merge branch 'branch-Level-10'

DKHB0207 committed:
Refactor taskList.txt

DKHB0207 committed:
build.gradle: Enable assertions in runtime

Assertions are disabled by default in Java applications. This commit
updates the build.gradle file to enable assertions, ensuring that
they are active during runtime.

Let's enable assertions to catch potential programming errors
during testing and development.

DKHB0207 committed:
storage: Add assertions to validate file paths and task parsing

Added assertions in the Storage class to ensure that file paths and task lists
are valid during file operations. These assertions help catch potential programming
errors early during development and maintain internal consistency in file handling.

- Assert that the file path is not null in the constructor.
- Assert that the file path exists before reading tasks from the file.
- Assert that the task list is not null before saving tasks.
- Assert that each task parsed from the file is valid (not null).

These changes improve code reliability by ensuring that file paths and task lists
are properly initialised and handled during storage operations.

DKHB0207 committed:
tasklist: Add assertions to validate tasks and task list state

Added assertions in the TaskList class to ensure that tasks and the task list
are in valid states before performing operations. These assertions help ensure
the integrity of the task list and catch potential errors early during development.

- Assert that tasks passed to the constructor are not null.
- Assert that tasks added to the task list are not null.
- Assert that the task list is not empty before deleting or retrieving tasks.
- Assert that tasks being filtered by keyword are not null.

These changes enhance the robustness of task management by validating inputs
and ensuring the task list is correctly maintained.

DKHB0207 committed:
commands: Add assertions to validate inputs and task list state

Added assertions in multiple command classes to validate inputs such as tasks
and keywords, as well as the task list state before performing operations.
These assertions help prevent null values and ensure internal consistency
during command execution.

- **AddCommand**:
  - Assert that the task to be added is not null in the constructor.
  - Assert that the task list and storage are not null before saving the task.

- **DeleteCommand**:
  - Assert that the task index is within valid bounds in the constructor before performing the delete operation.
  - Assert that the task to be deleted is not null before deleting the task.

- **FindCommand**:
  - Assert that the keyword is not null or empty.
  - Assert that the task list is not null before searching.

- **HelpCommand**:
  - Assert that the help instructions file exists

- **ListCommand**:
  - Assert that the task list is not null before listing tasks.

- **MarkCommand**:
  - Assert that the task list is not null before marking a task as done.
  - Assert that the task index is within valid bounds before performing the mark operation.

- **UnmarkCommand**:
  - Assert that the task list is not null before unmarking a task.
  - Assert that the task index is within valid bounds before performing the unmark operation.

These changes make the command execution more reliable by validating critical inputs
and ensuring tasks are properly handled during operations.

DKHB0207 committed:
Add assertions for Snipe class and Parser class

DKHB0207 committed:
Remove magic number to adhere to code quality standard

DKHB0207 committed:
Merge pull request #2 from DKHB0207/branch-A-Assertions

Add Assertions to Ensure Valid States in Key Classes

DKHB0207 committed:
Merge master into branch-A-CodeQuality

DKHB0207 committed:
Merge pull request #3 from DKHB0207/branch-A-CodeQuality

Improve Code Quality for Better Readability and Maintainability

DKHB0207 committed:
Initialize ArchiveList Storage in Snipe Class

- Added initialization of ArchiveList storage during the Snipe class setup.
- Modified the constructor of the Snipe class to instantiate and configure ArchiveList.
- Ensured ArchiveList is correctly loaded from persistent storage at startup.
- Updated relevant documentation and comments to reflect the new initialization process.
- Added error handling to manage potential failures during the initialization of ArchiveList.

These changes ensure that ArchiveList is ready for use when Snipe is instantiated, enhancing data management and ensuring archived items are correctly handled throughout the application's lifecycle.

DKHB0207 committed:
Refactor getResponse Method to Include ArchiveTasks and ArchiveStorage Parameters

- Modified the getResponse method to accept additional parameters: archiveTasks and archiveStorage.
- Updated the method signature to take in archiveTasks (TaskList) and archiveStorage (Storage) to support archiving functionality.
- Adjusted method calls across the codebase to pass the appropriate archive parameters, ensuring compatibility with the refactored method.
- Added necessary logic within getResponse to handle interactions with archived tasks, allowing the archiving of tasks while maintaining a clean main task list.
- Enhanced error handling and adjusted the exception management to incorporate archive operations without disrupting the main flow of the application.
- Updated unit tests and added new tests to validate the functionality of archiving items, ensuring that archived tasks are correctly saved and retrieved.
- Improved documentation within the code to clarify the role of archiveTasks and archiveStorage in the getResponse method, helping future developers understand the integration of the archiving feature.

**Context:**
- This refactoring is part of the C-Archive extension, which introduces the ability to archive tasks, allowing users to remove tasks from the active list while retaining a historical record.
- Archiving provides a way for users to start fresh without losing important past data, enhancing the flexibility and usability of the app.

**Why:**
- Enhances the getResponse method to support the archiving feature, aligning with the app's goal of managing tasks efficiently.
- Provides users with a way to manage task lists dynamically, keeping the active list focused and clutter-free while maintaining a record of past items.
- Lays the groundwork for future improvements and extensions related to task management and archival functionality.

DKHB0207 committed:
Implement C-Archive Extension to Support Task Archiving

- Introduced the Archive feature that enables users to archive tasks, allowing them to remove tasks from the active list while maintaining a record.
- Added functionality to archive tasks into a separate storage file, preserving task data outside the main task list.
- Updated the Snipe class to initialise ArchiveList and ArchiveStorage during startup, ensuring archived tasks are managed separately from active tasks.
- Modified the getResponse method to handle archive commands, providing users with options to archive tasks and view archived items.
- Added the ability to mark tasks as archived and save them to a designated archive file, making it easy for users to declutter their main task list.
- Updated UI prompts to inform users when tasks are archived, with distinct messages to differentiate archived lists from active lists.
- Implemented new command handling for archiving operations, allowing users to archive tasks directly through simple commands.
- Enhanced error handling and user feedback when interacting with archive functionality, ensuring smooth operation and clear communication.
- Improved documentation to guide users on how to use the archive feature, including instructions on archiving, viewing archived tasks, and starting fresh with a clean task list.

**Why:**
- The Archive feature provides a powerful way for users to manage their tasks efficiently by archiving old or completed tasks without losing important historical data.
- It helps users maintain a focused and organised active task list, enhancing overall productivity and user experience.
- The archiving capability lays the foundation for future task management enhancements, allowing for improved task lifecycle management within the application.

**Next Steps:**
- Explore additional archive-related features, such as restoring tasks from the archive and filtering archived items.
- Continue refining the user experience with feedback and enhancements to the archiving commands and UI prompts.

DKHB0207 committed:
Merge branch 'branch-C-Archive'

DKHB0207 committed:
Set the Stage Title to "Snipe" in Main Application Class

- Added `stage.setTitle("Snipe");` to the Main class to ensure the application window displays the product name in the title bar.
- This change improves the user interface by clearly identifying the application name, aligning with product presentation standards.
- The title bar now correctly reflects the product name, enhancing the overall appearance of the GUI and meeting documentation requirements for product screenshots.

**Why:**
- Setting the stage title ensures the product name is visible in the window's title bar, which is important for both user recognition and product branding.
- This change addresses the project requirement that the title bar should display the application name, ensuring screenshots accurately represent the product.

**Impact:**
- Users will now see "Snipe" in the title bar when the application is running, providing a clear and professional display of the application's identity.
- No other functional changes were made; this update solely affects the visual appearance of the application window.

DKHB0207 committed:
Add Functionality to Close Application on "bye" User Input

- Implemented a new feature that allows the application to close when the user inputs a specific command ("bye").
- Added the `delayCloseApplication()` method to the `MainWindow` class, which introduces a 1-second delay before closing the application to ensure that the "bye" message is displayed to the user.
- Integrated the `delayCloseApplication()` method into the `handleUserInput()` method, checking for specific input conditions that trigger the closure.
- Used `PauseTransition` to manage the delay, providing a smoother user experience by allowing the final message to be visible before the application exits.
- Updated the `handleUserInput()` method to add the delay before executing `Platform.exit()`, ensuring clean termination of the application.

**Why:**
- This enhancement improves user interaction by allowing the application to display a goodbye message before closing, enhancing the overall user experience.
- The delayed closure approach ensures the interface remains responsive and clear, even when the application is about to terminate.

DKHB0207 committed:
Add Ui.png screenshot to docs folder for product showcase

- Added a representative screenshot of the application's GUI named Ui.png to the docs folder.
- The screenshot captures the full GUI window, including the title bar displaying the product name "Snipe".
- This addition aligns with project requirements to showcase the application's interface on the product website.
- The file is correctly named Ui.png, ensuring consistency even if the original format was different.

**Why:**
- The screenshot provides a visual representation of the application, helping users and stakeholders understand the interface design and layout.
- It supports the User Guide and documentation by providing a clear view of what the application looks like in action.

DKHB0207 committed:
Update User Guide in docs/README.md

- Added a comprehensive User Guide to help users navigate the chatbot's features and commands.
- Documented key commands: `list`, `mark`, `unmark`, `todo`, `deadline` , `event` , `delete`, `find`, `archive`, `archives` and `help`, including descriptions, example usages, and expected outputs.
- Provided troubleshooting tips and contact information for user support.
- Included a feedback section to encourage user suggestions and improvements.

This update aims to improve user experience by offering clear, concise guidance on how to use the chatbot effectively.

DKHB0207 committed:
Refactor code to include comprehensive Javadocs documentation

- Added Javadocs for new archiveTask method in TaskList class
- Enhanced method documentation with @param, @return, @throws, and @implNote tags where applicable, ensuring clear understanding of method behaviors.
- Improved code readability and maintainability by aligning the documentation with coding standards and best practices.
- Included specific details on method purpose and implementation notes to guide developers in understanding and correctly using the methods.

This refactoring effort aims to enhance code quality by making it more self-explanatory, reducing the learning curve for new developers, and preventing misuse of methods due to a lack of clear guidance.

DKHB0207 committed:
Refactor code to remove hardcoded file paths and use relative directories

WHAT
- edited file path to remove hardcoded file path to a relative file path
- ensure file path is initialised before reading task list

WHY
- done to ensure jar file works regardless of where it is run

DKHB0207 committed:
fix: resolve file path bugs and rename shadowJar file

- Fixed issues with incorrect file paths that caused errors during execution.
- Renamed the shadowJar file to ensure consistent naming conventions and avoid conflicts.
- Updated related build configurations to reflect the new file name.

DKHB0207 committed:
fix: correct help instruction and add text constant in Ui class

- Fixed a bug in the help instruction that could not parse the help instruction information due to changes in file path for the text.
- Moved the help instruction text to a constant in the Ui class for better maintainability.
- Improved code readability and reduced redundancy by centralising the help text.

DKHB0207 committed:
Ui: add archive commands to help instruction in Ui class

- Updated the help instruction string in the Ui class to include 'archive' and 'archives' commands.
- 'archive': Archives the specified task, removing it from the active list but retaining it for reference.
- 'archives': Displays a list of all archived tasks.
- Improved command documentation within the application to keep users informed about all available commands.
- Ensured consistency in command descriptions for a clearer user experience.

DKHB0207 committed:
Rewrite help instruction list

- Cleaned up help instruction to be clearer
- Ensures user is able to clearly see what commands

DKHB0207 committed:
Adjust JavaDoc description of Ui class

- Ensure readability for user
- Clarify usage of constants in Ui class

DKHB0207 committed:
Add Javadoc documentation to Main class in snipe.gui

- Documented the purpose and functionality of the Main class
- Added Javadoc for the start method
- Included documentation for the Snipe instance
- Clarified parameters and exception handling in the start method

DKHB0207 committed:
Fix indentation of Javadoc comments

- Corrected indentation to improve readability
- Ensured consistent formatting across
- Purely formatting changes

DKHB0207 committed:
Fix indentation of Storage class

- Improve readability for the user
- Ensures consistency across project
